extensions/jaeger/runtime/src/main/java/io/quarkus/jaeger/runtime/QuarkusJaegerMetricsFactory.java

```diff
@@ -45,12 +46,17 @@ public void durationMicros(long time) {
 
     @Override
     public Gauge createGauge(final String name, final Map<String, String> tags) {
-        JaegerGauge gauge = registry.register(meta(name, MetricType.GAUGE), new JaegerGauge(), toTagArray(tags));
+        Tag[] tagArray = toTagArray(tags);
+        JaegerGauge gauge = (JaegerGauge) registry.getGauges().get(new MetricID(name, tagArray));
+        if (gauge == null) {
```

Comment:
After https://github.com/quarkusio/quarkus/issues/7018 we automatically drop and re-create all metric registries during shutdown before a hot reload, so it should not happen anymore that the gauge is still registered. Are you sure we still need this? That would mean the registry dropping is wrong?!

Reply:
I've seen an exception during hot reload on this. I'll try again.

Reply:
No, this is still required. When I revert this change, I get this on hot reload:

```
org.jboss.resteasy.spi.UnhandledException: java.lang.RuntimeException: No reflection exceptions should be thrown unless there is a fundamental error in your code set up.
        at org.jboss.resteasy.core.ExceptionHandler.handleException(ExceptionHandler.java:381)
        at org.jboss.resteasy.core.SynchronousDispatcher.writeException(SynchronousDispatcher.java:216)
        at org.jboss.resteasy.core.SynchronousDispatcher.invoke(SynchronousDispatcher.java:515)
        at org.jboss.resteasy.core.SynchronousDispatcher.lambda$invoke$4(SynchronousDispatcher.java:259)
        at org.jboss.resteasy.core.SynchronousDispatcher.lambda$preprocess$0(SynchronousDispatcher.java:160)
        at org.jboss.resteasy.core.interception.jaxrs.PreMatchContainerRequestContext.filter(PreMatchContainerRequestContext.java:362)
        at org.jboss.resteasy.core.SynchronousDispatcher.preprocess(SynchronousDispatcher.java:163)
        at org.jboss.resteasy.core.SynchronousDispatcher.invoke(SynchronousDispatcher.java:245)
        at io.quarkus.resteasy.runtime.standalone.RequestDispatcher.service(RequestDispatcher.java:73)
        at io.quarkus.resteasy.runtime.standalone.VertxRequestHandler.dispatch(VertxRequestHandler.java:122)
        at io.quarkus.resteasy.runtime.standalone.VertxRequestHandler.access$000(VertxRequestHandler.java:36)
        at io.quarkus.resteasy.runtime.standalone.VertxRequestHandler$1.run(VertxRequestHandler.java:87)
        at io.quarkus.runtime.CleanableExecutor$CleaningRunnable.run(CleanableExecutor.java:231)
        at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
        at java.util.concurrent.FutureTask.run(FutureTask.java:266)
        at org.jboss.threads.ContextClassLoaderSavingRunnable.run(ContextClassLoaderSavingRunnable.java:35)
        at org.jboss.threads.EnhancedQueueExecutor.safeRun(EnhancedQueueExecutor.java:2027)
        at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.doRunTask(EnhancedQueueExecutor.java:1551)
        at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1442)
        at org.jboss.threads.DelegatingRunnable.run(DelegatingRunnable.java:29)
        at org.jboss.threads.ThreadLocalResettingRunnable.run(ThreadLocalResettingRunnable.java:29)
        at java.lang.Thread.run(Thread.java:748)
        at org.jboss.threads.JBossThread.run(JBossThread.java:479)
Caused by: java.lang.RuntimeException: No reflection exceptions should be thrown unless there is a fundamental error in your code set up.
        at io.jaegertracing.internal.metrics.Metrics.createMetrics(Metrics.java:76)
        at io.jaegertracing.internal.metrics.Metrics.<init>(Metrics.java:32)
        at io.jaegertracing.internal.metrics.Metrics.<init>(Metrics.java:28)
        at io.jaegertracing.Configuration.getTracerBuilder(Configuration.java:221)
        at io.quarkus.jaeger.runtime.QuarkusJaegerTracer.tracer(QuarkusJaegerTracer.java:40)
        at io.quarkus.jaeger.runtime.QuarkusJaegerTracer.buildSpan(QuarkusJaegerTracer.java:59)
        at io.opentracing.util.GlobalTracer.buildSpan(GlobalTracer.java:133)
        at io.quarkus.smallrye.opentracing.runtime.TracerProducer_ProducerMethod_tracer_96dadb3d6afa0cccadfe742c3e06ad433737c844_ClientProxy.buildSpan(TracerProducer_ProducerMethod_tracer_96dadb3d6afa0cccadfe742c3e06ad433737c844_ClientProxy.zig:238)
        at io.opentracing.contrib.jaxrs2.server.ServerTracingFilter.filter(ServerTracingFilter.java:69)
        at org.jboss.resteasy.core.interception.jaxrs.PreMatchContainerRequestContext.filter(PreMatchContainerRequestContext.java:310)
        at org.jboss.resteasy.core.ResourceMethodInvoker.invokeOnTarget(ResourceMethodInvoker.java:439)
        at org.jboss.resteasy.core.ResourceMethodInvoker.invoke(ResourceMethodInvoker.java:400)
        at org.jboss.resteasy.core.ResourceMethodInvoker.invoke(ResourceMethodInvoker.java:374)
        at org.jboss.resteasy.core.ResourceMethodInvoker.invoke(ResourceMethodInvoker.java:67)
        at org.jboss.resteasy.core.SynchronousDispatcher.invoke(SynchronousDispatcher.java:488)
        ... 20 more
Caused by: java.lang.IllegalArgumentException: A metric with metricID MetricID{name='jaeger_tracer_reporter_queue_length', tags=[]} already exists
        at io.smallrye.metrics.MetricsRegistryImpl.register(MetricsRegistryImpl.java:131)
        at io.quarkus.jaeger.runtime.QuarkusJaegerMetricsFactory.createGauge(QuarkusJaegerMetricsFactory.java:48)
        at io.jaegertracing.internal.metrics.Metrics.createMetrics(Metrics.java:66)
        ... 34 more
```

Reply:
@jmartisk Any idea? Thanks! :-)

Reply:
That's fishy, I'll try to reproduce this and let you know

Reply:
@Ladicek I can't reproduce that - I created a simple app which uses tracing and metrics, called one traced method, triggered a reload, called that traced method again, the `jaeger_tracer_reporter_queue_length` gauge was re-registered correctly and no errors occurred.
I used Quarkus built from this PR branch, but with this one change reverted. Any other specific thing that I should try? :)

Reply:
Oh well. I've just verified this still occurs, even if I rebase this PR on top of current `master`. I'll put together a detailed reproducer.

Reply:
@jmartisk Here's a reproducer with detailed instructions: https://github.com/Ladicek/quarkus-jaeger-metrics-hot-reload-issue

Reply:
@Ladicek  Ha! Thanks for the reproducer. I think I understand why this happens now and why it didn't happen with my reproducer: your application does **not** have a dependency on the `quarkus-smallrye-metrics` module, so the metrics processor doesn't run and doesn't register the task to clean up registries during reload.

But what now? Is it correct that Jaeger registers metrics even if the application does not declare a dependency on `quarkus-smallrye-metrics`? In other extensions, we generally try to make their dependency on metrics optional, so metrics are activated only if the user explicitly adds the  `quarkus-smallrye-metrics` dependency.

Reply:
And if we can't make changes to the Jaeger extension such that metrics become optional (and we can run without them completely), then I suppose we should make the Jaeger module depend on `quarkus-smallrye-metrics-deployment` in order to make sure that the metrics processor is always applied and can do proper cleanup

Reply:
It also happens that if the Metrics extension is not present (as in your reproducer), then metrics are not being exposed on the `/metrics` endpoint, so they are quite useless.

Reply:
Now that is ... interesting. `quarkus-jaeger` depends on `smallrye-metrics` and `microprofile-metrics-api`, but not on `quarkus-smallrye-metrics`. That is ... wrong, for sure.

Reply:
Or perhaps that was done to make metrics semi-optional...

Reply:
@jmartisk What's the way to detect if metrics are enabled? I guess `capabilities.isCapabilityPresent(Capabilities.METRICS)`? If so, then we could decide on that and add either `NoopMetricsFactory` or `QuarkusJaegerMetricsFactory`. And probably make the dependencies on `smallrye-metrics` and `microprofile-metrics-api` `<scope>provided</scope>`, because they are only present to be able to compile `QuarkusJaegerMetricsFactory`. Does that sound about right? If so, I can check if that works and submit a PR.

Reply:
@Ladicek yes, checking for the capability is ok.
I think the correct way is have an `optional` dependency on `quarkus-smallrye-metrics`, like Agroal does it:
https://github.com/quarkusio/quarkus/blob/master/extensions/agroal/runtime/pom.xml#L44

Reply:
And what would be absolutely optimal and consistent with most other extensions, would be that the metrics from the extension are disabled by default (**even if** the metrics dependency is present) and have to be enabled by a config property `quarkus.jaeger.metrics.enabled`. See https://quarkus.io/guides/writing-extensions#extension-metrics

Even more optimal would be using the Metrics SPI (mentioned in the linked document) to register metrics eagerly, but if you feel that's too much for now, I guess we can leave that for later :)

Reply:
Seriously? :open_mouth: :grin:

Reply:
I mean, I would probably be able to do all that, but that would make this PR pretty large. I just wanted to fix the trace context propagation, found a dev mode issue, included a fix which was tiny, and now this :-/

Here's my proposal: I already have finished the conditional inclusion/exclusion of metrics based on the capability. I have also changed `quarkus-jaeger` to no longer depend on `smallrye-metrics` and `microprofile-metrics-api`; instead, it uses an `optional` dependency on `quarkus-smallrye-metrics` as you suggest. It seems to work (didn't try native :grin:). What's is _not_ done is addressing your 2nd comment, the one starting with "And what would be absolutely optimal" -- I'd leave that for future. Agree?

Reply:
Sure, that's clearly not the goal of this PR. 
I created https://github.com/quarkusio/quarkus/issues/8502 for that

Reply:
Nice, thanks! I'll update this PR in a few minutes.

Reply:
PR rebased and updated per discussion above.